core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -368,6 +352,100 @@ public void write(MarshalledEntry<? extends K, ? extends V> marshalledEntry) {
+      if (meta != null)
+         buf.put(meta.getBuf(), meta.getOffset(), meta.getLength());
+      buf.flip();
+      return buf;
+   }
+
+   private FileEntry allocateAndCreateFileEntry(MarshalledEntry entry,
+                                                org.infinispan.commons.io.ByteBuffer key,
+                                                org.infinispan.commons.io.ByteBuffer data,
+                                                org.infinispan.commons.io.ByteBuffer meta) {
+      int metadataLength = meta == null ? 0 : meta.getLength();
+      int len = KEY_POS + key.getLength() + data.getLength() + metadataLength;
+      FileEntry newEntry = allocate(len);
+      long expiryTime = meta != null ? entry.getMetadata().expiryTime() : -1;
+      return new FileEntry(newEntry, key.getLength(), data.getLength(), metadataLength, expiryTime, entry);
+   }
+
+   private FileEntry allocateAndCreateFileEntry(MarshalledEntry entry) {
+      org.infinispan.commons.io.ByteBuffer key = entry.getKeyBytes();
+      org.infinispan.commons.io.ByteBuffer data = entry.getValueBytes();
+      org.infinispan.commons.io.ByteBuffer meta = entry.getMetadataBytes();
+
+      return allocateAndCreateFileEntry(entry, key, data, meta);
+   }
+
+   @Override
+   public void writeBatch(Collection<MarshalledEntry<? extends K, ? extends V>> marshalledEntries) {
+      resizeLock.readLock().lock();
+      try {
+         List<FileEntry> fileEntries = marshalledEntries.stream().map(this::allocateAndCreateFileEntry)
+               .collect(Collectors.toList());
+         fileEntries.sort((o1, o2) -> {
+            long offset1 = o1.offset;
+            long offset2 = o2.offset;
+            return offset1 < offset2 ? -1 : offset1 == offset2 ? 0 : 1;
+         });
+         for (FileEntry entry : fileEntries) {
+            MarshalledEntry<K, V> marshalledEntry = entry.marshalledEntry;
+            ByteBuffer buf = createByteBuffer(entry);
+            channel.write(buf, entry.offset);
+
+            if (trace) log.tracef("Wrote entry %s:%d at %d:%d", marshalledEntry.getKey(), entry.size, entry.offset, entry.size);
+            FileEntry oldEntry = entries.put(marshalledEntry.getKey(), entry);
+            if (oldEntry == null)
+               oldEntry = evict();
+            free(oldEntry);
+         }
+      } catch (Exception e) {
+         throw new PersistenceException(e);
+      } finally {
+         resizeLock.readLock().unlock();
+
+      }
+   }
+
+   @Override
+   public void deleteBatch(Collection<Object> keys) {
+      ArrayList<KeyValuePair<K, FileEntry>> keysToLoad = new ArrayList<>(keys.size());
+      resizeLock.readLock().lock();
+      try {
+         synchronized (entries) {
+            for (Object key : keys) {
+               FileEntry entry = entries.get(key);
+               if (entry != null)
+                  keysToLoad.add(new KeyValuePair<>((K) key, entry));
+            }
+         }
+
+         keysToLoad.sort((o1, o2) -> {
```

Comment:
Okay I defer to your judgement @danberindei as I am not as familiar with how it would perform in scenarios where the writes are scattered in different positions among the file.